pkg/helios/cas/cas.go

```diff
@@ -0,0 +1,398 @@
+	s.mutex.Lock()
+	s.cache[hashKey] = make([]byte, len(content))
+	copy(s.cache[hashKey], content)
+	s.mutex.Unlock()
+	
+	// Cache the hex key for future lookups
+	s.keyMutex.Lock()
+	s.keyCache[string(digest)] = hashKey
+	s.keyMutex.Unlock()
+
+	// Store persistently to disk - async for performance or skip in memory mode
+	if !s.memoryMode {
+		filePath := filepath.Join(s.storePath, hashKey)
+		
+		// Critical: Add to WaitGroup BEFORE attempting to send to channel
+		// to prevent race condition where Done() is called before Add()
+		s.wg.Add(1)
+		
+		// Use async writes for better performance
+		select {
+		case s.writeQueue <- writeOp{filePath: filePath, content: content}:
+			// Successfully queued for background write
+		default:
+			// Queue full, must call Done() since we already called Add()
+			s.wg.Done()
+			// Write synchronously as fallback
+			if err := os.WriteFile(filePath, content, 0644); err != nil {
+				return hash, fmt.Errorf("failed to write content to disk: %w", err)
+			}
+		}
+	}
+
+	return hash, nil
+}
+
+// StoreBatch processes multiple content items in a single optimized operation
+// Designed for VST commit scenarios requiring <70μs for 50 files
+func (s *BLAKE3Store) StoreBatch(contents [][]byte) ([]types.Hash, error) {
+	hashes := make([]types.Hash, len(contents))
+	
+	// Pre-allocate maps to avoid repeated allocations
+	hashKeys := make([]string, len(contents))
+	
+	// Process all hashes first (CPU-bound, can be optimized)
+	for i, content := range contents {
+		// Get hasher from pool for zero allocation
+		hasher := s.hasherPool.Get().(*blake3.Hasher)
+		hasher.Write(content)
+		digest := hasher.Sum(nil)
+		hasher.Reset()
+		s.hasherPool.Put(hasher)
+		
+		hashes[i] = types.Hash{
+			Algorithm: types.BLAKE3,
+			Digest:    digest,
+		}
+		hashKeys[i] = fmt.Sprintf("%x", digest)
+	}
+	
+	// Batch cache operations (single lock acquisition)
+	s.mutex.Lock()
+	for i, content := range contents {
+		hashKey := hashKeys[i]
+		if _, exists := s.cache[hashKey]; !exists {
+			s.cache[hashKey] = make([]byte, len(content))
+			copy(s.cache[hashKey], content)
+		}
+	}
+	s.mutex.Unlock()
+	
+	// Batch key cache operations
+	s.keyMutex.Lock()
+	for i, hash := range hashes {
+		digestStr := string(hash.Digest)
+		s.keyCache[digestStr] = hashKeys[i]
+	}
+	s.keyMutex.Unlock()
+	
+	// Handle persistence based on mode
+	if !s.memoryMode {
+		// Queue all writes together for better batching
+		for i, content := range contents {
+			filePath := filepath.Join(s.storePath, hashKeys[i])
+			
+			// Critical: Add to WaitGroup BEFORE attempting to send to channel
+			s.wg.Add(1)
+			
+			select {
+			case s.writeQueue <- writeOp{filePath: filePath, content: content}:
+				// Successfully queued
+			default:
+				// Queue full, must call Done() since we already called Add()
+				s.wg.Done()
+				// Fallback to sync write
+				os.WriteFile(filePath, content, 0644) // Ignore error in batch mode
```

Comment:
Silent error: the fallback sync write in batch mode ignores errors. Suggest routing errors into errorQueue (like backgroundWriter), or accumulating and returning a combined error so callers can react. Silent data loss is risky.